File: typescript/nodes/list_urls.ts
import { next, Nextable } from "../ambler.ts";
import { State } from "../state.ts";

type ListEdges = { onSuccess: Nextable<State> };

export function listUrls(edges: ListEdges) {
  return (state: State) => {
    console.log("\n--- URLs ---");
    state.urls.forEach((url) => console.log(url));
    console.log("------------");
    return next(edges.onSuccess, state);
  };
}

File: python/nodes/list_urls.py
from typing import TypedDict

from ambler import Next, Nextable
from state import State


class ListEdges(TypedDict):
    on_success: Nextable


def list_urls(edges: ListEdges) -> Nextable:
    async def _node(state: State) -> Next[State]:
        print("\n--- URLs ---")
        for url in state['urls']:
            print(url)
        print("------------")
        return Next(edges['on_success'], state)

    return _node
